lib/statsd/instrument/client.rb

```diff
@@ -0,0 +1,142 @@
+  #
+  # @param sample_rate [Float] (default: nil)
+  # @param tags [Hash, Array] (default: nil)
+  # @return StatsD::Instrument::Datagram
+  def increment(name, value = 1, sample_rate: nil, tags: nil)
+    sample(sample_rate, datagram_builder.c(name, value, sample_rate, tags))
+  end
+
+  # Emits a timing metric.
+  # @param name [String] The name of the metric.
+  # @param value [Numeric] The timing to record in milliseconds
+  # @param sample_rate [Float] (default: nil)
+  # @param tags [Hash, Array] (default: nil)
+  # @return StatsD::Instrument::Datagram
+  def measure(name, value = nil, sample_rate: nil, tags: nil)
+    sample(sample_rate, datagram_builder.ms(name, value, sample_rate, tags))
+  end
+
+  # Emits a gauge metric.
+  # @param name [String] The name of the metric.
+  # @param value [Numeric] The gauged value
+  # @param sample_rate [Float] (default: nil)
+  # @param tags [Hash, Array] (default: nil)
+  # @return StatsD::Instrument::Datagram
+  def gauge(name, value, sample_rate: nil, tags: nil)
+    sample(sample_rate, datagram_builder.g(name, value, sample_rate, tags))
+  end
+
+  # Emits a set metric, which counts unique values.
+  # @param name [String] The name of the metric.
+  # @param value [Numeric, String] (default: 1) The value to record
+  # @param sample_rate [Float] (default: nil)
+  # @param tags [Hash, Array] (default: nil)
+  # @return StatsD::Instrument::Datagram
+  def set(name, value, sample_rate: nil, tags: nil)
+    sample(sample_rate, datagram_builder.s(name, value, sample_rate, tags))
+  end
+
+  # Instantiates a new StatsD client that uses the settings of the current client,
+  # except for the provided settings.
+  def with_options(
+    sink: nil,
+    prefix: nil,
+    sampling_disabled: nil,
+    default_sample_rate: nil,
+    default_tags: nil,
+    datagram_builder_class: nil
+  )
+    client = clone_with_options(sink: sink, prefix: prefix, sampling_disabled: sampling_disabled,
+      default_sample_rate: default_sample_rate, default_tags: default_tags,
+      datagram_builder_class: datagram_builder_class)
+
+    yield(client)
+  end
+
+  def clone_with_options(
+    sink: nil,
+    prefix: nil,
+    sampling_disabled: nil,
+    default_sample_rate: nil,
+    default_tags: nil,
+    datagram_builder_class: nil
+  )
+    self.class.new(
+      sink: sink || @sink,
+      prefix: prefix || @prefix,
+      sampling_disabled: sampling_disabled || @sampling_disabled,
+      default_sample_rate: default_sample_rate || @default_sample_rate,
+      default_tags: default_tags || @default_tags,
+      datagram_builder_class: datagram_builder_class || @datagram_builder_class,
+    )
+  end
+
+  def capture_sink
+    StatsD::Instrument::CaptureSink.new(parent: @sink)
+  end
+
+  def with_capture_sink(capture_sink)
+    @sink = capture_sink
+    yield
+    @sink.datagrams
+  ensure
+    @sink = @sink.parent
+  end
+
+  def capture(&block)
+    with_capture_sink(capture_sink, &block)
```

Comment:
I don't quite see the value of `with_capture_sink` being extracted, `capture` end up being a one liner, and only caller of `with_capture_sink`. Seems like both method could be merged without any increase in complexity.

Reply:
I can leave this out, but we need this to implement `assert_statsd_`-methods correctly. See https://github.com/Shopify/statsd-instrument/pull/166